src/ucp/stream/stream_recv.c

```diff
@@ -0,0 +1,131 @@
+                                          stream_queue);
+    *length = rdesc->length;
+    ret = (ucs_status_ptr_t)((uintptr_t)(rdesc + 1) + rdesc->hdr_len);
+
+out:
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+
+    return ret;
+}
+
+UCS_PROFILE_FUNC_VOID(ucp_stream_data_release, (ep, data),
+                      ucp_ep_h ep, void *data)
+{
+    ucp_recv_desc_t *rdesc;
+
+    /* TODO: make const header offset */
+    rdesc = (ucp_recv_desc_t *)((uintptr_t)(data -
+                                            sizeof(ucp_stream_am_hdr_t))) - 1;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&ep->worker->mt_lock);
+
+    if (ucs_unlikely(rdesc->flags & UCP_RECV_DESC_FLAG_UCT_DESC)) {
+        uct_iface_release_desc((char*)rdesc - sizeof(ucp_eager_sync_hdr_t));
+    } else {
+        ucs_mpool_put_inline(rdesc);
+    }
+
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&ep->worker->mt_lock);
+}
+
+static ucs_status_t ucp_stream_am_handler(void *arg, void *data, size_t length,
+                                          unsigned am_flags)
+{
+    const size_t            hdr_len = sizeof(ucp_stream_am_hdr_t);
+    ucp_worker_h            worker  = arg;
+    ucp_ep_h                ep;
+    ucp_stream_am_hdr_t     *hdr;
+    ucp_recv_desc_t         *rdesc;
+    size_t                  recv_len;
+    khiter_t                hash_it;
+    ucs_status_t            status;
+
+    UCP_THREAD_CS_ENTER_CONDITIONAL(&worker->mt_lock);
+
+    ucs_assert(length >= hdr_len);
+    recv_len = length - hdr_len;
+
+    /* TODO: merge common code with TAG unexpected */
+    if (ucs_unlikely(am_flags & UCT_CB_PARAM_FLAG_DESC)) {
+        /* slowpath */
+        rdesc        = (ucp_recv_desc_t *)data - 1;
+        rdesc->flags = UCP_RECV_DESC_FLAG_UCT_DESC;
+        status       = UCS_INPROGRESS;
+    } else {
+        rdesc = (ucp_recv_desc_t*)ucs_mpool_get_inline(&worker->am_mp);
+        if (rdesc == NULL) {
+            ucs_error("ucp recv descriptor is not allocated");
+            status = UCS_ERR_NO_MEMORY;
+            goto out;
+        }
+
+        rdesc->flags = 0;
+        memcpy(rdesc + 1, data, length);
+        status = UCS_OK;
+    }
+
+    rdesc->length  = recv_len;
+    rdesc->hdr_len = hdr_len;
+    hdr            = data;
+
+    hash_it = kh_get(ucp_worker_ep_hash, &worker->ep_hash, hdr->uuid);
+    if (hash_it != kh_end(&worker->ep_hash)) {
+        ep = kh_value(&worker->ep_hash, hash_it);
+        ucs_queue_push(&ep->stream_data, &rdesc->stream_queue);
+    } else {
+        ucs_error("ep is not found by uuid: %lu", hdr->uuid);
+        status = UCS_ERR_SOME_CONNECTS_FAILED;
+    }
+
+out:
+    UCP_THREAD_CS_EXIT_CONDITIONAL(&worker->mt_lock);
+    return status;
+}
+
+static void ucp_stream_am_dump(ucp_worker_h worker, uct_am_trace_type_t type,
+                               uint8_t id, const void *data, size_t length,
+                               char *buffer, size_t max)
+{
+    /* TODO: */
```

Comment:
can you pls fill this as part of this PR

Reply:
Ok